### PR TITLE
[DPE-3946] Add actions

### PR DESCRIPTION
**Pull request by welpaolo**

### CODE REVIEW

**Review by deusebio — APPROVED** [review]
well done @welpaolo ! I also like that we use the peer relation to store this properly, and we have extended the different classes consistently with the new pattern. Well done!

I only have a couple of non-blocking comment to be discussed and - if relevant - to be addressed on dedicated items. 

1. I'm wondering whether we should allow to add/remove more than one property at once. I'm thinking of 

    1. Allowing to remove multiple keys at the same time, assuning the key parameter is comma separated
    2. Allowing another (otpional) parameter in the add-config maybe called `raw` or something that can be easily fed with:

```
juju run hub/leader add-config raw=$(cat spark.properties)
```

2. Store these properties as juju secrets instead of directly in the databag. Similar as right now that we store these properties in a K8s secret, we should probably expose these in proper secrets since they may contain sensible informations. 

If we don't have obvious reasons against them, we could create separate tickets and address/finalize the design in the ticket itself, but I wanted to start suggesting them.

File: src/managers/integration_hub.py [hunk 1]
```diff
@@ -35,6 +41,8 @@ def _ssl_enabled(endpoint: str | None) -> str:
     def _s3_conf(self) -> dict[str, str]:
         if s3 := self.s3:
             return {
+                "spark.hadoop.fs.s3a.path.style.access": "true",
```
Comment (theoctober19th) [inline src/managers/integration_hub.py:44]: +1 for adding this.

File: src/managers/integration_hub.py [hunk 2]
```diff
@@ -35,6 +41,8 @@ def _ssl_enabled(endpoint: str | None) -> str:
...
+                "spark.eventLog.enabled": "true",
```
Comment (theoctober19th) [inline src/managers/integration_hub.py:45]: Do we really want this? Wouldn't this enable event logs automatically on S3 relation? In my view, this configuration is not quite related with S3 integration.

Comment (theoctober19th) [reply src/managers/integration_hub.py:45]: Also, if we are adding this configuration, I think `spark.eventLog.dir` would accompany this as well.

File: tests/integration/test_charm.py [hunk 3]
```diff
@@ -214,6 +226,103 @@ async def test_build_and_deploy(ops_test: OpsTest, charm_versions):
...
+    res = await run_action(ops_test, "add-config", {"conf": "a=b"})
+    assert res["return-code"] == 0
+    # wait for active status
+    await ops_test.model.wait_for_idle(
+        apps=[APP_NAME],
+        status="active",
+        timeout=1000,
+    )
+    logger.info(f"add-config action result: {res}")
+    sleep(5)
```
Comment (theoctober19th) [inline tests/integration/test_charm.py:265]: Same as below.

**Review by theoctober19th — APPROVED** [review]
I have a few minor comments, otherwise it looks good to me.

**Comment by theoctober19th** [comment]
> well done @welpaolo ! I also like that we use the peer relation to store this properly, and we have extended the different classes consistently with the new pattern. Well done!
> 
> I only have a couple of non-blocking comment to be discussed and - if relevant - to be addressed on dedicated items.
> 
> 1. I'm wondering whether we should allow to add/remove more than one property at once. I'm thinking of
>    
>    1. Allowing to remove multiple keys at the same time, assuning the key parameter is comma separated
>    2. Allowing another (otpional) parameter in the add-config maybe called `raw` or something that can be easily fed with:
> 
> ```
> juju run hub/leader add-config raw=$(cat spark.properties)
> ```
> 
> 2. Store these properties as juju secrets instead of directly in the databag. Similar as right now that we store these properties in a K8s secret, we should probably expose these in proper secrets since they may contain sensible informations.
> 
> If we don't have obvious reasons against them, we could create separate Jira tickets and address/finalize the design in the ticket itself, but I wanted to start suggesting them.

I agree that a feature to allow for multiple configurations to be added in at the same time would definitely be a desired feature. I prefer multiple args (just like `--conf a=b --conf c=d` as we did in spark8t if possible. If not, ability to read the properties from a file (kinda similar to `--properties-file` in spark8t)as well as stdin would be better IMO. If that's the case, one can simply do:

```
spark-client.service-account-registry get-config --username u --namespace n | juju run integration-hub-k8s/0 add-config

OR

spark-client.service-account-registry get-config --username u --namespace n > /tmp/propfile
juju run integration-hub-k8s/0 add-config --propfile /tmp/propfile
```

This can be done as a separate task and doesn't necessarily block this to be merged, however.

**Comment by deusebio** [comment]
About 1, I'm not sure if that would be supported by juju run actions, but I agree and I would also prefer a multiple args syntax if possible (but really, I don't know/think)

File: tests/integration/test_charm.py [hunk 4]
```diff
@@ -214,6 +226,103 @@ async def test_build_and_deploy(ops_test: OpsTest, charm_versions):
...
+    # clear config
+    res = await run_action(ops_test, "clear-config", {})
+    assert res["return-code"] == 0
+    # wait for active status
+    await ops_test.model.wait_for_idle(
+        apps=[APP_NAME],
+        status="active",
+        timeout=1000,
+    )
+    sleep(5)
```
[the remark this answers is not in the document]
Comment (welpaolo) [reply tests/integration/test_charm.py:317]: Yes, it is intentional. On CI, the secrets update could be slow so I just added sleep to ensure that secrets are updated.

File: tests/integration/test_charm.py (hunk 3 above)
Comment (welpaolo) [reply tests/integration/test_charm.py:265]: Same as above

**Comment by welpaolo** [comment]
I will open a ticket for the support multiple configurations at once

File: src/constants.py [hunk 5]
```diff
@@ -8,6 +8,7 @@
 
 CONTAINER = "integration-hub"
 INTEGRATION_HUB_LABEL = "app.kubernetes.io/managed-by=integration-hub"
+PEER = "hub"
```
[the remark this answers is not in the document]
Comment (welpaolo) [reply src/constants.py]: Make it more meaningful

**Pull request merged**